zmanim/astronomical_calendar.py

```diff
@@ -0,0 +1,103 @@
+        if geo_location is None:
+            geo_location = GeoLocation.GMT()
+        if date is None:
+            date = datetime.today()
+        if calculator is None:
+            calculator = NOAACalculator()
+        self.geo_location = geo_location
+        self.date = date
+        self.astronomical_calculator = calculator
+
+    def sunrise(self) -> datetime:
+        return self.__date_time_from_time_of_day(self.utc_sunrise(self.GEOMETRIC_ZENITH), 'sunrise')
+
+    def sea_level_sunrise(self) -> datetime:
+        return self.sunrise_offset_by_degrees(self.GEOMETRIC_ZENITH)
+
+    def sunrise_offset_by_degrees(self, offset_zenith: float) -> datetime:
+        return self.__date_time_from_time_of_day(self.utc_sea_level_sunrise(offset_zenith), 'sunrise')
+
+    def sunset(self) -> datetime:
+        return self.__date_time_from_time_of_day(self.utc_sunset(self.GEOMETRIC_ZENITH), 'sunset')
+
+    def sea_level_sunset(self) -> datetime:
+        return self.sunset_offset_by_degrees(self.GEOMETRIC_ZENITH)
+
+    def sunset_offset_by_degrees(self, offset_zenith: float) -> datetime:
+        return self.__date_time_from_time_of_day(self.utc_sea_level_sunset(offset_zenith), 'sunset')
+
+    def utc_sunrise(self, zenith: float) -> float:
+        return self.astronomical_calculator.utc_sunrise(self.__adjusted_date(), self.geo_location, zenith, adjust_for_elevation=True)
+
+    def utc_sea_level_sunrise(self, zenith: float) -> float:
+        return self.astronomical_calculator.utc_sunrise(self.__adjusted_date(), self.geo_location, zenith, adjust_for_elevation=False)
+
+    def utc_sunset(self, zenith: float) -> float:
+        return self.astronomical_calculator.utc_sunset(self.__adjusted_date(), self.geo_location, zenith, adjust_for_elevation=True)
+
+    def utc_sea_level_sunset(self, zenith: float) -> float:
+        return self.astronomical_calculator.utc_sunset(self.__adjusted_date(), self.geo_location, zenith, adjust_for_elevation=False)
+
+    def temporal_hour(self, sunrise: datetime = __sentinel, sunset: datetime = __sentinel) -> float:
+        if sunrise == self.__sentinel:
+            sunrise = self.sea_level_sunrise()
+        if sunset == self.__sentinel:
+            sunset = self.sea_level_sunset()
+
+        if sunset is None or sunrise is None:
+            return None
+
+        daytime_hours = float((sunset - sunrise).total_seconds() / 3600.0)
+        return (daytime_hours / 12) * self.HOUR_MILLIS
+
+    def sun_transit(self) -> datetime:
+        sunrise = self.sea_level_sunrise()
+        sunset = self.sea_level_sunset()
+        if sunrise is None or sunset is None:
+            return None
+        noon_hour = (self.temporal_hour(sunrise, sunset) / self.HOUR_MILLIS) * 6.0
+        return sunrise + timedelta(noon_hour / 24.0)
+
+    def __date_time_from_time_of_day(self, time_of_day: float, mode: str) -> datetime:
```

Comment:
This returns an `Optional[datetime]`, based on line 80

Reply:
Also, since it appears that `time_of_day` accepts None, the arg type should be `Optional[float]` as well.